src/NUnitFramework/framework/Interfaces/TNode.cs

```diff
@@ -318,15 +319,69 @@ private static NodeList ApplySelection(NodeList nodeList, string xpath)
                 : resultNodes;
         }
 
-        private static readonly Regex InvalidXmlCharactersRegex = new Regex("[^\u0009\u000a\u000d\u0020-\ufffd]|([\ud800-\udbff](?![\udc00-\udfff]))|((?<![\ud800-\udbff])[\udc00-\udfff])", RegexOptions.Compiled);
         private static string EscapeInvalidXmlCharacters(string str)
         {
             if (str == null) return null;
 
-            // Based on the XML spec http://www.w3.org/TR/xml/#charsets
-            // For detailed explanation of the regex see http://mnaoumov.wordpress.com/2014/06/15/escaping-invalid-xml-unicode-characters/
+            StringBuilder builder = null;
+            for (int i = 0; i < str.Length; i++)
+            {
+                char c = str[i];
+                if(c > 0x20 && c < 0x7F)
+                {
+                    // ASCII characters - break quickly for these
+                    if (builder != null)
+                        builder.Append(c);
+                }
+                // From the XML specification: http://www.w3.org/TR/xml/#charsets
+                // Char ::= #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF]
+                // Any Unicode character, excluding the surrogate blocks, FFFE, and FFFF.
+                else if (!(0x0 <= c && c <= 0x8) &&
+                    c != 0xB &&
+                    c != 0xC &&
+                    !(0xE <= c && c <= 0x1F) &&
+                    !(0x7F <= c && c <= 0x84) &&
+                    !(0x86 <= c && c <= 0x9F) &&
+                    !(0xD800 <= c && c <= 0xDFFF) &&
+                    c != 0xFFFE &&
+                    c != 0xFFFF)
+                {
+                    if (builder != null)
+                        builder.Append(c);
+                }
+                // Also check if the char is actually a high/low surogate pair of two characters.
+                // If it is, then it is a valid XML character (from above based on the surrogate blocks).
+                else if (char.IsHighSurrogate(c) &&
+                    i + 1 != str.Length &&
+                    char.IsLowSurrogate(str[i + 1]))
+                {
+                    if (builder != null)
+                    {
+                        builder.Append(c);
+                        builder.Append(str[i + 1]);
+                    }
+                    i++;
+                }
+                else
+                {
+                    // We keep the builder null so that we don't allocate a string
+                    // when doing this conversion until we encounter a unicode character.
+                    // Then, we allocate the rest of the string and escape the invalid
+                    // character.
+                    if (builder == null)
+                    {
+                        builder = new StringBuilder();
```

Comment:
My last comment/question (sorry about not thinking about this earlier): will there be any noticeable difference if we use one of the other overloads of the constructor to minimize expanding the `StringBuilder`? (especially when `str` is a larger string). We could do something like
```suggestion
                        builder = new StringBuilder(str.Length);
```
or also avoiding the loop (hope I get the boundaries right)
```suggestion
                        builder = new StringBuilder(str.Substring(0, index)), str.Length);
```

Reply:
I considered it, but since we don't know how large the string will be, I didn't figure there would be much of a performance benefit, as whatever capacity we passed in would have to immediately be changed anyway.